lib/awful/ewmh.lua

```diff
@@ -202,12 +145,63 @@ function ewmh.urgent(c, urgent)
     end
 end
 
+-- Map the state to the action name
+local context_mapper = {
+    maximized_vertical   = "maximize_vertically",
+    maximized_horizontal = "maximize_horizontally",
+    fullscreen           = "maximize"
+}
+
+--- Move and resize the client.
+--
+-- This is the default geometry request handler.
+--
+-- @tparam client c The client
+-- @tparam string context The context
+-- @tparam[opt={}] table hints The hints to pass to the handler
+function ewmh.geometry(c, context, hints)
+    context = context or ""
+
+    local original_context = context
+
+    -- Now, map it to something useful
+    context = context_mapper[context] or context
+
+    local props = util.table.clone(hints or {}, false)
+    props.store_geometry = props.store_geometry==nil and true or props.store_geometry
+
+    -- If it is a known placement function, then apply it, otherwise let
+    -- other potential handler resize the client (like in-layout resize or
+    -- floating client resize)
+    if aplace[context] then
+
+        -- Check if it correspond to a boolean property
+        local state = c[original_context]
+
+        -- If the property is boolean and it correspond to the undo operation,
+        -- restore the stored geometry.
+        if state == false then
+            aplace.restore(c,{context=context})
+            return
+        end
+
+        local honor_default = original_context ~= "fullscreen"
+
+        if props.honor_workarea == nil then
+            props.honor_workarea = honor_default
+        end
+
+        aplace[context](c, props)
+    elseif context == "fullscreen" then
+        props.honor_workarea = false
```

Comment:
Doesn't this also need to check `c.fullscreen`, like in the previous case (`local state = c[original_context]`)?
Put differently: Does unfullscreening work?